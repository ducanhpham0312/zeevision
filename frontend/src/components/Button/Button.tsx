import React, { ButtonHTMLAttributes } from "react";
import { twMerge } from "tailwind-merge";
import { tv, type VariantProps } from "tailwind-variants";

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariant> {
  width?: number;
  children?: React.ReactNode;
  active?: boolean;
  label?: string;
  helperText?: string;
}

export function Button({
  children,
  width,
  active,
  className,
  variant,
  size,
  fullWidth,
  label,
  helperText,
  ...props
}: ButtonProps) {
  return (
    <button
      {...props}
      className={twMerge(
        buttonVariant({ variant, size, fullWidth }),
        className,
      )}
      style={{ width: `${width}px` }}
    >
      {helperText ? (
        <div className="absolute -left-3 top-[50%] -translate-x-[100%] -translate-y-[50%] scale-50 rounded bg-[#202020] px-2 py-1 text-white opacity-0 transition group-hover:scale-100 group-hover:opacity-100">
          <p className="whitespace-nowrap text-sm">{helperText}</p>
        </div>
      ) : null}
      <div
        className={`transition-width absolute bottom-0 left-0 w-full transform rounded bg-accent duration-100`}
        style={{ height: active ? "4px" : "0px" }}
      />
      <span>{label || children}</span>
    </button>
  );
}

const buttonVariant = tv({
  base: "group hover:shadow-lg transition relative box-border text-accent hover:bg-hover active:bg-active rounded",
  variants: {
    variant: {
      primary: "text-white bg-accent hover:bg-accent/90 active:bg-accent/80",
      secondary: "bg-second-accent",
    },
    size: {
      compact: "px-2 py-1",
      standard: "px-3 py-2",
      large: "px-4 py-4",
    },
    fullWidth: {
      true: "w-full",
    },
  },
  defaultVariants: {
    size: "standard",
  },
});
