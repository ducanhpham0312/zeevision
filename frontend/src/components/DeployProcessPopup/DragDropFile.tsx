import { Modal } from "@mui/base/Modal";

import { useEffect } from "react";
import { useDragDrop } from "../../hooks/useDragDrop";
import { readFileToString } from "../../utils/readFileToString";
import { useUIStore } from "../../contexts/useUIStore";

interface DragDropFileProps {
  /**
   * Callback function that is triggered when a file is dropped.
   *
   * @param fileContent - The content of the dropped file as a string.
   */
  onFileDropped: (fileContent: string) => void;
}

export function DragDropFile({ onFileDropped }: DragDropFileProps) {
  const { dragging, file } = useDragDrop();
  const { setSnackbarContent } = useUIStore();

  useEffect(() => {
    const setFile = async () => {
      if (file) {
        if (file.name.endsWith(".bpmn")) {
          onFileDropped(await readFileToString(file));
        } else {
          setSnackbarContent({
            title: "Invalid file type",
            message: "The file was not .bpmn file",
            type: "error",
          });
        }
      }
    };

    setFile();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [file]);

  return (
    <Modal
      className="pointer-events-none fixed inset-0 z-50 flex items-center justify-center"
      open={dragging}
      aria-labelledby="simple-modal-title"
      aria-describedby="simple-modal-description"
    >
      <div className="flex h-[200px] w-[400px] flex-col items-center justify-center rounded border-[3px] border-text">
        <p id="simple-modal-description">Drop files here to upload!</p>
        <em>(Only *.bmpn files will be accepted)</em>
      </div>
    </Modal>
  );
}
