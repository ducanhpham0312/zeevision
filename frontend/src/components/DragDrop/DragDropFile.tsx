import { Modal } from "@mui/base/Modal";
import { Box, styled } from "@mui/system";

import React, { useEffect } from "react";
import { PRIMARY } from "../../theme/palette";
import { useDragDrop } from "../../hooks/useDragDrop";
import { ReadBpmnFileToString } from "../../utils/ReadBpmnFileToString";

interface DragDropFileProps {
  onFileDropped: (file: string) => void;
}

export const DragDropFile: React.FC<DragDropFileProps> = ({
  onFileDropped,
}) => {
  const { dragging, file } = useDragDrop();

  useEffect(() => {
    ReadBpmnFileToString(file, onFileDropped);

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [file]);

  return (
    <StyledModal
      sx={{ pointerEvents: "none" }}
      open={dragging}
      aria-labelledby="simple-modal-title"
      aria-describedby="simple-modal-description"
    >
      <StyledBox>
        <p id="simple-modal-description">Drop files here to upload!</p>
        <em>(Only *.bmpn files will be accepted)</em>
      </StyledBox>
    </StyledModal>
  );
};

const StyledModal = styled(Modal)`
  position: fixed;
  z-index: 1300;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const StyledBox = styled(Box)`
  width: 400px;
  height: 200px;
  border: 3px solid ${PRIMARY[600]};
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: #e8f8ff;
  cursor: pointer;
  margin: 16px;

  p {
    color: ${PRIMARY[300]};
    font-size: 16px;
    text-align: center;
  }
`;
