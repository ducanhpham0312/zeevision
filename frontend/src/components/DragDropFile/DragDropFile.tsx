import { Modal } from "@mui/base/Modal";
import { Box, styled } from "@mui/system";

import React, { useEffect } from "react";
import { PRIMARY } from "../../theme/palette";
import { useDragDrop } from "../../hooks/useDragDrop";
import { readFileToString } from "../../utils/readFileToString";
import { useUIStore } from "../../contexts/useUIStore";

interface DragDropFileProps {
  /**
   * Callback function that is triggered when a file is dropped.
   *
   * @param fileContent - The content of the dropped file as a string.
   */
  onFileDropped: (file: string) => void;
}

export const DragDropFile: React.FC<DragDropFileProps> = ({
  onFileDropped,
}) => {
  const { dragging, file } = useDragDrop();
  const { setSnackbarContent } = useUIStore();

  useEffect(() => {
    const setFile = async () => {
      if (file) {
        if (file.name.endsWith(".bpmn")) {
          onFileDropped(await readFileToString(file));
        } else {
          setSnackbarContent({
            title: "Invalid file type",
            message: "The file was not .bpmn file",
            type: "error",
          });
        }
      }
    };

    setFile();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [file]);

  return (
    <StyledModal
      sx={{ pointerEvents: "none" }}
      open={dragging}
      aria-labelledby="simple-modal-title"
      aria-describedby="simple-modal-description"
    >
      <StyledBox>
        <p id="simple-modal-description">Drop files here to upload!</p>
        <em>(Only *.bmpn files will be accepted)</em>
      </StyledBox>
    </StyledModal>
  );
};

export const StyledModal = styled(Modal)`
  position: fixed;
  z-index: 1300;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const StyledBox = styled(Box)`
  width: 400px;
  height: 200px;
  border: 3px solid ${PRIMARY[600]};
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: #e8f8ff;
  cursor: pointer;
  margin: 16px;

  p {
    color: ${PRIMARY[300]};
    font-size: 16px;
    text-align: center;
  }
`;
