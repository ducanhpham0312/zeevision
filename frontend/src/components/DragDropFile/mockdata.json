{
  "file": {
    "content": "Test File DragnDrop",
    "name": "test.bpmn",
    "type": "text/plain"
  }
}
