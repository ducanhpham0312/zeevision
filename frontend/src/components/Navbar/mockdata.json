[
  {
    "name": "PROCESSES",
    "path": "/processes"
  },
  {
    "name": "INSTANCES",
    "path": "/instances"
  },
  {
    "name": "INCIDENTS",
    "path": "/incidents"
  },
  {
    "name": "JOBS",
    "path": "/jobs"
  },
  {
    "name": "MESSAGES",
    "path": "/messages"
  },
  {
    "name": "ERRORS",
    "path": "/errors"
  }
]
