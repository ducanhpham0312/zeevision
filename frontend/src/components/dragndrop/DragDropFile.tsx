import { Modal } from "@mui/base/Modal";
import { Box, styled } from "@mui/system";
import { useCallback, useEffect, useState } from "react";
import { useDropzone } from "react-dropzone";

import React from "react";

interface DragDropFileProps {}

export const DragDropFile: React.FC<DragDropFileProps> = () => {
  const [isDragActive, setIsDragActive] = useState(false);
  const [file, setFile] = useState<File | null>(null);

  const onDrop = useCallback((acceptedFiles: File[]) => {
    // Only accept bpmn files
    if (acceptedFiles.length > 0 && acceptedFiles[0].name!.endsWith(".bpmn")) {
      setFile(acceptedFiles[0]);
    }
    console.log(file);
    setIsDragActive(false);
  }, []);

  useEffect(() => {
    // Function to handle drag over window
    const handleDragOver = (e: Event) => {
      e.preventDefault();
      setIsDragActive(true);
    };

    // Function to handle drag leave window
    const handleDragLeave = () => {
      setIsDragActive(false);
    };

    // Adding event listeners
    window.addEventListener("dragover", handleDragOver);
    window.addEventListener("dragleave", handleDragLeave);
    window.addEventListener("drop", handleDragLeave);

    // Removing event listeners on component unmount
    return () => {
      window.removeEventListener("dragover", handleDragOver);
      window.removeEventListener("dragleave", handleDragLeave);
      window.removeEventListener("drop", handleDragLeave);
    };
  }, []);

  const { getRootProps, getInputProps } = useDropzone({
    onDrop,
    noClick: true, // Prevent file dialog on click
    noKeyboard: true, // Prevent keyboard navigation
  });

  return (
    <div
      style={{
        display: "none",
        position: "fixed",
        height: "100vh",
        top: 0,
        width: "100vw",
        boxSizing: "border-box",
        border: "10px solid red",
      }}
    >
      {isDragActive && (
        <DropzoneArea {...getRootProps()}>
          <input {...getInputProps()} />
          <StyledModal
            open={isDragActive}
            onClose={() => setIsDragActive(false)}
            aria-labelledby="simple-modal-title"
            aria-describedby="simple-modal-description"
          >
            <StyledBox>
              <p id="simple-modal-description">Drop files here to upload!</p>
              <em>(Only *.bmpn files will be accepted)</em>
            </StyledBox>
          </StyledModal>
        </DropzoneArea>
      )}
    </div>
  );
};

const StyledModal = styled(Modal)`
  position: fixed;
  z-index: 1300;
  inset: 0;
  display: flex;
  align-items: center;
  justify-content: center;
`;
const DropzoneArea = styled(Box)`
  width: 100%;
  height: 100vh;
  border: 2px dashed lightgray;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
`;
const StyledBox = styled(Box)`
  width: 400px;
  height: 200px;
  border: 3px dashed #97c8eb;
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: #e8f8ff;
  cursor: pointer;
  margin: 16px;
  &:hover {
    border-color: #4194e0;
  }
  p {
    color: #4194e0;
    font-size: 16px;
    text-align: center;
  }
`;
