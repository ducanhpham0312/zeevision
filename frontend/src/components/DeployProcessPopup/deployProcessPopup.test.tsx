import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { DeployProcessPopup } from ".";
import * as mockdata from "./mockdata.json";
import { Button } from "../Button";
import { DragDropFile } from "./DragDropFile";

const fileContent = mockdata.file.content;
const fileName = mockdata.file.name;
const fileType = mockdata.file.type;

jest.mock("../../hooks/useDragDrop", () => {
  const useDragDropMock = jest.fn(() => ({
    dragging: true,
    file: new File([fileContent], fileName, { type: fileType }),
  }));

  return { useDragDrop: useDragDropMock };
  // Mocking a problematic component
});
jest.mock("../BpmnViewer/BpmnViewer.tsx", () => {
  return {
    __esModule: true,
    BpmnViewer: jest.fn(() => <div>BpmnViewer Mock</div>),
    ResponsiveBpmnViewer: jest.fn(() => <div>BpmnViewer Mock</div>),
  };
});

describe("DeployProcessPopup Component", () => {
  let handleOpen: jest.Mock<void, []>, handleClose: jest.Mock<void, []>;

  beforeEach(() => {
    handleOpen = jest.fn();
    handleClose = jest.fn();
  });

  it("renders the popup correctly when a file is dragged over", () => {
    const { baseElement } = render(
      <DeployProcessPopup
        isPopUpOpen={true}
        onOpenPopUp={handleOpen}
        onClosePopUp={handleClose}
      />,
    );

    const file = new File([fileContent], fileName, { type: fileType });

    fireEvent.dragOver(document);
    fireEvent.drop(document, { dataTransfer: { files: [file] } });

    expect(baseElement).toMatchSnapshot();
  });

  it("should display the modal when 'Deploy a Process' button is clicked", () => {
    render(
      <>
        <Button onClick={handleOpen}>Deploy a Process</Button>
        <DeployProcessPopup
          isPopUpOpen={false}
          onOpenPopUp={handleOpen}
          onClosePopUp={handleClose}
        />
      </>,
    );
    fireEvent.click(screen.getByText("Deploy a Process"));
    expect(handleOpen).toHaveBeenCalledTimes(1);
  });

  it("should trigger the file input when 'Deploy a file' button in Popup is clicked", () => {
    render(
      <DeployProcessPopup
        isPopUpOpen={true}
        onOpenPopUp={handleOpen}
        onClosePopUp={handleClose}
      />,
    );

    const fileInput = screen.getByTestId("file-input");
    const mockFileInputClick = jest.spyOn(fileInput, "click");

    fireEvent.click(screen.getByText("Upload a file (.bpmn)"));

    expect(mockFileInputClick).toHaveBeenCalled();

    // Clean up after the test
    mockFileInputClick.mockRestore();
  });

  it("renders Bpmn view when a file is uploaded via 'Deploy a file' button in Popup", async () => {
    render(
      <DeployProcessPopup
        isPopUpOpen={true}
        onOpenPopUp={handleOpen}
        onClosePopUp={handleClose}
      />,
    );

    // Mocking the file input change event to simulate a file upload
    const fileInput = screen.getByTestId("file-input");
    const file = new File([fileContent], fileName, { type: fileType });
    fireEvent.change(fileInput, { target: { files: [file] } });

    // Wait for the mock BpmnViewer to be in the document
    await waitFor(() => {
      expect(screen.getByText("BpmnViewer Mock")).toBeInTheDocument();
    });
  });

  it("renders Bpmn view when a file is uploaded via drag and drop in window", async () => {
    render(
      <DeployProcessPopup
        isPopUpOpen={true}
        onOpenPopUp={handleOpen}
        onClosePopUp={handleClose}
      />,
    );

    const file = new File([fileContent], fileName, { type: fileType });

    fireEvent.dragOver(document);
    fireEvent.drop(document, { dataTransfer: { files: [file] } });

    await waitFor(() => {
      const bpmnViewerElement = screen.getByText("BpmnViewer Mock");
      expect(bpmnViewerElement).toBeInTheDocument();
    });
  });

  it("renders Bpmn view when a file is uploaded via drag and drop in Popup", async () => {
    render(
      <DeployProcessPopup
        isPopUpOpen={true}
        onOpenPopUp={handleOpen}
        onClosePopUp={handleClose}
      />,
    );

    const file = new File([fileContent], fileName, { type: fileType });

    // Drop zone in Popup
    const dropZone = screen.getByText("Or drag a bpmn file here");

    fireEvent.dragOver(dropZone);
    fireEvent.drop(dropZone, { dataTransfer: { files: [file] } });

    await waitFor(() => {
      const bpmnViewerElement = screen.getByText("BpmnViewer Mock");
      expect(bpmnViewerElement).toBeInTheDocument();
    });
  });

  it("calls onClose when close button is clicked", () => {
    render(
      <DeployProcessPopup
        isPopUpOpen={true}
        onOpenPopUp={handleOpen}
        onClosePopUp={handleClose}
      />,
    );

    fireEvent.click(screen.getByText("Cancel"));
    expect(handleClose).toHaveBeenCalledTimes(1);
  });
});

describe("DragDropFile", () => {
  let onFileDropped: jest.Mock<void, []>;

  beforeEach(() => {
    onFileDropped = jest.fn();
  });

  it("shows the modal when dragging", () => {
    render(<DragDropFile onFileDropped={onFileDropped} />);

    expect(screen.getByText("Drop files here to upload!")).toBeInTheDocument();
  });

  it("calls onFileDropped when a file is dropped", async () => {
    render(<DragDropFile onFileDropped={onFileDropped} />);

    const file = new File([fileContent], fileName, { type: fileType });
    fireEvent.drop(document, { dataTransfer: { files: [file] } });

    await waitFor(() => {
      expect(onFileDropped).toHaveBeenCalledWith(fileContent);
    });
  });
});
