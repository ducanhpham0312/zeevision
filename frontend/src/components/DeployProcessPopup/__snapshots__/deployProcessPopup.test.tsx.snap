// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeployProcessPopup Component renders the popup correctly when a file is dragged over 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div />
  </div>
  <div
    aria-hidden="true"
    class="MuiModal-root css-1lnsth0"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiModal-backdrop css-1jyxheb"
      style="opacity: 0;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      style="opacity: 0;"
      tabindex="-1"
    >
      <div
        class="css-1mxqldv"
      >
        <div
          class="css-rjt43j"
        >
          <p>
            Deploy a process
          </p>
          <button
            class="MuiButton-root css-39vfrt"
            type="button"
            variant="text"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
        </div>
        <div
          class="css-12e8uj5"
          style="overflow: hidden;"
        >
          <div>
            <button
              class="MuiButton-root css-1lm9n5u"
              type="button"
              variant="outlined"
            >
              Deploy a file (.bpmn)
            </button>
            <p>
              Or drag a bpmn file here
            </p>
            <input
              accept=".bpmn"
              data-testid="file-input"
              style="display: none;"
              type="file"
            />
          </div>
          <div />
        </div>
        <div
          class="css-1csjh8m"
        >
          <button
            class="MuiButton-root css-1lm9n5u"
            type="button"
            variant="outlined"
          >
            Cancel
          </button>
          <button
            class="MuiButton-root css-11j4va6"
            type="button"
            variant="contained"
          >
            Deploy process
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
  <div
    aria-describedby="simple-modal-description"
    aria-labelledby="simple-modal-title"
    class="MuiModal-root css-y1nyvh"
    role="presentation"
  >
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiBox-root css-idn9ht"
      tabindex="-1"
    >
      <p
        id="simple-modal-description"
      >
        Drop files here to upload!
      </p>
      <em>
        (Only *.bmpn files will be accepted)
      </em>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
